### PR TITLE
TCT integration

### PR DESCRIPTION
**Pull request by plaidfinch**
This PR comprises a variety of changes along the way to integrating the TCT into the rest of the Penumbra codebase. They can be seen as falling into several distinct categories:

- internal improvements to the TCT discovered during the course of integration,
- a full executable specification and property test suite for the TCT developed to gain confidence before integration,
- interface improvements to the TCT discovered during the course of integration, and
- changes to various parts of the Penumbra codebase outside of the TCT to integrate it and remove the `incrementalmerkletree` crate, notably:
  - breaking changes to the proto formats for transparent proofs,
  - drop-in integration into the `crypto` and `transaction` crates to replace equivalent parts of the previous NCT,
  - changes to the `wallet` crate to support scanning compact blocks into the new TCT design: this is currently _not_ optimized using the TCT's unique summarizing features, but it is written to automatically roll into new epochs based on the block height,
  - fixes to `pcli` to update code written while this PR was pending to use the new crate and its API when necessary in transaction construction, and
  - changes and fixes to `pd` to integrate the TCT into the new component architecture for the shielded pool.

I should note that I have not yet tested the combination of these changes to `pd` and `pcli` to ensure that they produce a coherent result. That should happen before this merges.

### CODE REVIEW

File: Cargo.toml
Comment (hdevalence) [inline Cargo.toml:24]: This is changing the entire workspace configuration to do more expensive builds, regardless of whether or not the expensive tests are actually going to be run.

Could we just do `cargo test --release` instead when running the slow, long-running property tests?

File: crypto/src/note.rs
Comment (hdevalence) [inline crypto/src/note.rs:352]: Having two copies of the (logically) same type in two different crates is pretty awkward, although I guess it's not the worst thing for now.

File: crypto/src/proofs/transparent.rs
Comment (hdevalence) [inline crypto/src/proofs/transparent.rs:609]: This name is kind of confusing out of context. I think it would be better if the structures from the `tct` crate had `Tree` in the name, like `EternityTree`, `EpochTree`, `BlockTree`, etc.

File: crypto/src/proofs/transparent.rs
Comment (hdevalence) [inline crypto/src/proofs/transparent.rs:610]: Rather than re-exporting the inner `Witness::Keep` and `Witness::Forget` variants from the crate root, then calling them `tct::Keep`, it would be clearer if the calling code just imported `Witness` and wrote
```
nct.insert(Witness::Keep, note_commitment);
```

File: pd/src/components/shielded_pool.rs
Comment (hdevalence) [inline pd/src/components/shielded_pool.rs:278]: This function shouldn't return a `Result`, because we cannot fail to execute a transaction once we've accepted it with `check_tx_stateful`.

Instead, we need to check that the `BlockTree` is not full in the `check_tx_stateful` method, and then `.expect` (with a reference to that check) that the `BlockTree` is not full in `execute_tx`.

File: pd/src/components/shielded_pool.rs
Comment (hdevalence) [inline pd/src/components/shielded_pool.rs:306]: This code special-cases `0`, and skips inserting a new block after processing the genesis state.  This means that the genesis allocations will end up in the same tree as any notes created in the first block, which seems wrong.

I think the root cause of this problem is that this function was originally just supposed to write the block, but ended up containing the `end_block` implementation.

File: pd/src/components/shielded_pool.rs
Comment (hdevalence) [inline pd/src/components/shielded_pool.rs:312]: This assertion is wrong, because it does not compute the height correctly (it will fail as soon as the first epoch ends, halting the chain).

File: pd/src/components/shielded_pool.rs
Comment (hdevalence) [inline pd/src/components/shielded_pool.rs:304]: It's a bit unintuitive that the way to finish a block or epoch is to insert an additional, empty structure.  I think it would be better to have `finish_block` and `finish_epoch` methods (even if they have this implementation internally).

File: tct/Cargo.toml
Comment (hdevalence) [inline tct/Cargo.toml:24]: We should not include `sqlx` here, since we've just finished standardizing on using our canonical, proto-based serialization for persistent storage.

File: tct/src/epoch.rs
Comment (hdevalence) [inline tct/src/epoch.rs:213]: Why are we panicking here instead of just returning `None` as before?

File: tct/src/epoch.rs
Comment (hdevalence) [inline tct/src/epoch.rs:553]: We don't need this code, because we don't use Postgres any more.

File: tct/src/internal/hash.rs
Comment (hdevalence) [inline tct/src/internal/hash.rs:171]: This code should also not be added, since we don't use Postgres any more.

File: tct/src/internal/hash/cache.rs
Comment (hdevalence) [inline tct/src/internal/hash/cache.rs:29]: We should absolutely not be doing an `unsafe impl Sync` in our Merkle tree implementation.  It is critical that the Merkle tree work correctly, and having to reason about synchronization behaviour on top of basic functional correctness adds a huge amount of risk for very little gain.

File: tct/src/internal/hash/cache.rs
Comment (hdevalence) [inline tct/src/internal/hash/cache.rs:61]: The fact that accidentally leaving a `#[derive(Debug)]` on the struct (or adding one in the future, or ...) could break the soundness of the code we use to verify note inclusion seems like a very clear sign that using `unsafe` here is a bad way to allocate implementation risk.

File: tct/src/internal/hash/fast_hash.rs
Comment (hdevalence) [inline tct/src/internal/hash/fast_hash.rs:4]: We shouldn't make this change in this PR, because it is already far too large, but wouldn't this be a non-problem if the tree were generic over the leaf type, like the `incrementalmerkletree` API?

File: wallet/src/state.rs
Comment (hdevalence) [inline wallet/src/state.rs:816]: The errors should be `Sync` to avoid this type of problem.

**Review by hdevalence — CHANGES_REQUESTED** [review]
We should not use unsafe `Sync` impls in our Merkle tree implementation.  There are some other minor API design points, but this is the major blocker.  It would be better to just use a `parking_lot` mutex.

We should also run the code to see whether it works correctly before merging -- currently, it panics on the first epoch transition, because of https://github.com/penumbra-zone/penumbra/pull/617#discussion_r851674520 , but even after that, the client can't submit any transactions, because it doesn't compute the same NCT root.

File: pd/src/components/shielded_pool.rs
Comment (hdevalence) [reply pd/src/components/shielded_pool.rs]: Fixed in https://github.com/penumbra-zone/penumbra/pull/617/commits/cda7931722303f4894e333fad8e1d29b65a65305

File: pd/src/components/shielded_pool.rs
Comment (hdevalence) [reply pd/src/components/shielded_pool.rs]: Fixed in https://github.com/penumbra-zone/penumbra/pull/617/commits/cda7931722303f4894e333fad8e1d29b65a65305

**Comment by hdevalence** [comment]
Fixed the chain halt and block 0 / block 1 merge in cda7931, but transactions still fail because the submitted anchor is not valid.  I noticed that 
```
Apr 17 00:58:37.965 DEBUG abci:EndBlock{height=344}:end_block:shielded_pool:write_block: pd::components::shielded_pool: writing anchor height=0 anchor=Root(d132c20a)
```
The `height` field is obviously wrong there, but there's also something wrong with the `Debug` implementation for `Root` that's causing it to print a truncated hash.  Looking into the code, the `Root`'s `Debug` implementation is derived from that of `Hash`, which is doing something mysterious:
```rust
impl Debug for Hash {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
        let mut bytes = Vec::with_capacity(4 * 8);
        self.0.write(&mut bytes).unwrap();
        write!(f, "{}", hex::encode(&bytes[3 * 8 + 4..]))
    }
}
```

**Comment by hdevalence** [comment]
Why does the NCT root change every fourth block, even when there are no new notes?
```
Apr 17 02:05:22.896 DEBUG scan_block{height=16}: penumbra_wallet::state: finished scanning block self.last_block_height=16 root=Root(43570e07)
Apr 17 02:05:22.899 DEBUG scan_block{height=17}: penumbra_wallet::state: starting block scan outputs_len=0
Apr 17 02:05:22.899 DEBUG scan_block{height=17}: penumbra_wallet::state: inserting block tree
Apr 17 02:05:22.899 DEBUG scan_block{height=17}: penumbra_wallet::state: finished scanning block self.last_block_height=17 root=Root(43570e07)
Apr 17 02:05:22.899 DEBUG scan_block{height=18}: penumbra_wallet::state: starting block scan outputs_len=0
Apr 17 02:05:22.899 DEBUG scan_block{height=18}: penumbra_wallet::state: inserting block tree
Apr 17 02:05:22.899 DEBUG scan_block{height=18}: penumbra_wallet::state: finished scanning block self.last_block_height=18 root=Root(43570e07)
Apr 17 02:05:22.899 DEBUG scan_block{height=19}: penumbra_wallet::state: starting block scan outputs_len=0
Apr 17 02:05:22.899 DEBUG scan_block{height=19}: penumbra_wallet::state: inserting block tree
Apr 17 02:05:22.899 DEBUG scan_block{height=19}: penumbra_wallet::state: finished scanning block self.last_block_height=19 root=Root(43570e07)
Apr 17 02:05:22.899 DEBUG scan_block{height=20}: penumbra_wallet::state: starting block scan outputs_len=0
Apr 17 02:05:22.899 DEBUG scan_block{height=20}: penumbra_wallet::state: inserting block tree
Apr 17 02:05:22.899 DEBUG scan_block{height=20}: penumbra_wallet::state: finished scanning block self.last_block_height=20 root=Root(bd866301)
Apr 17 02:05:22.899 DEBUG scan_block{height=21}: penumbra_wallet::state: starting block scan outputs_len=0
Apr 17 02:05:22.899 DEBUG scan_block{height=21}: penumbra_wallet::state: inserting block tree
Apr 17 02:05:22.899 DEBUG scan_block{height=21}: penumbra_wallet::state: finished scanning block self.last_block_height=21 root=Root(bd866301)
Apr 17 02:05:22.899 DEBUG scan_block{height=22}: penumbra_wallet::state: starting block scan outputs_len=0
Apr 17 02:05:22.899 DEBUG scan_block{height=22}: penumbra_wallet::state: inserting block tree
Apr 17 02:05:22.899 DEBUG scan_block{height=22}: penumbra_wallet::state: finished scanning block self.last_block_height=22 root=Root(bd866301)
Apr 17 02:05:22.899 DEBUG scan_block{height=23}: penumbra_wallet::state: starting block scan outputs_len=0
Apr 17 02:05:22.899 DEBUG scan_block{height=23}: penumbra_wallet::state: inserting block tree
Apr 17 02:05:22.899 DEBUG scan_block{height=23}: penumbra_wallet::state: finished scanning block self.last_block_height=23 root=Root(bd866301)
Apr 17 02:05:22.899 DEBUG scan_block{height=24}: penumbra_wallet::state: starting block scan outputs_len=0
Apr 17 02:05:22.899 DEBUG scan_block{height=24}: penumbra_wallet::state: inserting block tree
Apr 17 02:05:22.899 DEBUG scan_block{height=24}: penumbra_wallet::state: finished scanning block self.last_block_height=24 root=Root(a1e1ac04)
```

**Comment by hdevalence** [comment]
After some further rearrangement to fix NCT root desyncing between the client and the server, it no longer falls out of sync immediately.  However, it looks like there's something wrong with the way that adding a complete block tree works, because it ends up different from adding an empty one and then appending to it.

On the `pd` side, we have
```
Apr 17 02:47:23.263 DEBUG abci:DeliverTx{txid="53acd053b9a070de23be87a4243b2da165c72afda856a560cb954e9441167f1a"}:execute_tx:shielded_pool:add_note: pd::components::shielded_pool: appending to NCT in component commitment=note::Commitment(e4c09957dc8b720e909b47a84d6737bdc464b9f5053f6331b291aaa21e7cb907) position=Position(Eternity { epoch: 1, block: 19, commitment: 0 })
```
so the note commitment is inserted at position `1/19/0`, and then in `end_block`, we have
```
Apr 17 02:47:23.264 DEBUG abci:EndBlock{height=60}:end_block:shielded_pool: pd::components::shielded_pool: inserting block tree for next block
```
and then
```
Apr 17 02:47:23.265 DEBUG abci:EndBlock{height=60}:end_block:shielded_pool:write_block: pd::components::shielded_pool: position=Position(Eternity { epoch: 1, block: 20, commitment: 0 })
Apr 17 02:47:23.265 DEBUG abci:EndBlock{height=60}:end_block:shielded_pool:write_block: pd::components::shielded_pool: writing anchor height=60 anchor=Root(5362b403)
```

On the `pcli` side, we have
```
Apr 17 02:47:29.305 DEBUG scan_block{height=60}: penumbra_wallet::state: position=Position(Eternity { epoch: 1, block: 19, commitment: 0 })
Apr 17 02:47:29.305 DEBUG scan_block{height=60}: penumbra_wallet::state: starting block scan outputs_len=1
Apr 17 02:47:29.305 DEBUG scan_block{height=60}: penumbra_wallet::state: found note while scanning note_commitment=note::Commitment(e4c09957dc8b720e909b47a84d6737bdc464b9f5053f6331b291aaa21e7cb907) position=Position(Eternity { epoch: 1, block: 19, commitment: 0 }) note=Note { value: Value { amount: 100000000000, asset_id: passet1984fctenw8m2fpl8a9wzguzp7j34d7vravryuhft808nyt9fdggqxmanqm }, note_blinding: Fp256(BigInteger256([1925076761441888379, 10327062928397409300, 16562428622360418953, 14453853609863477])), diversifier: Diversifier(e21c212d0d11002a7f5ab5), transmission_key: decaf377_ka::Public(025df1c0885c83a1f9f4370ec0e3edec336ffc513bd80611c15b0c9df6d0c901), transmission_key_s: Fp256(BigInteger256([15986969656633463360, 6500317334198110268, 4067783070624381186, 1271587261249631793])) }
Apr 17 02:47:29.306 DEBUG scan_block{height=60}: penumbra_wallet::state: inserting block tree
Apr 17 02:47:29.306 DEBUG scan_block{height=60}: penumbra_wallet::state: found nullifier for submitted spend note, marking it as spent value=Value { amount: 100000000000, asset_id: passet1984fctenw8m2fpl8a9wzguzp7j34d7vravryuhft808nyt9fdggqxmanqm } nullifier=Nullifier(e292ad67f1a986df243a3d5ce7ec385c4fadfc6ce5ff333f221efc760f8f0604)
Apr 17 02:47:29.306 DEBUG scan_block{height=60}: penumbra_wallet::state: finished scanning block self.last_block_height=60 root=Root(77f69003) position=Position(Eternity { epoch: 1, block: 20, commitment: 1 })
```
so although the note is inserted in the same `1/19/0` position, we end up with a different position and a different root hash.

**Comment by hdevalence** [comment]
Adding more tracing suggests there's a bug in `insert_block`, since inserting a block tree to advance the position doesn't do so:
```
Apr 17 03:29:07.985 DEBUG abci:EndBlock{height=40}:end_block:shielded_pool: pd::components::shielded_pool: position=Position(Eternity { epoch: 1, block: 0, commitment: 0 })
Apr 17 03:29:07.985 DEBUG abci:EndBlock{height=40}:end_block:shielded_pool: pd::components::shielded_pool: inserting block tree for next block
Apr 17 03:29:07.985 DEBUG abci:EndBlock{height=40}:end_block:shielded_pool: pd::components::shielded_pool: position=Position(Eternity { epoch: 1, block: 0, commitment: 0 })
```

File: pd/src/components/shielded_pool.rs
Comment (hdevalence) [reply pd/src/components/shielded_pool.rs:329]: The `Component` API doesn't make it clear that this is a problem, because there's no distinction in what *kind* of `Result`s are returned.  Discussion of a potential fix: https://github.com/penumbra-zone/penumbra/issues/632

**Comment by hdevalence** [comment]
From discussion with @plaidfinch, since this PR is very large but not immediately mergeable, the current plan is to try to separate this branch into two parts, one that only touches the `tct` code, and another that touches the rest of the workspace.

File: tct/src/internal/hash/cache.rs
Comment (plaidfinch) [reply tct/src/internal/hash/cache.rs:29]: This is now resolved in the tct-integration-tct-only branch.

File: tct/src/internal/hash/cache.rs
Comment (plaidfinch) [reply tct/src/internal/hash/cache.rs:61]: This is now resolved in the tct-integration-tct-only branch.

File: tct/Cargo.toml
Comment (plaidfinch) [reply tct/Cargo.toml:24]: Sqlx has been removed in the tct-integration-tct-only branch.

File: tct/src/epoch.rs
Comment (plaidfinch) [reply tct/src/epoch.rs:553]: Removed in tct-integration-tct-only.

File: tct/src/internal/hash.rs
Comment (plaidfinch) [reply tct/src/internal/hash.rs:171]: Removed in tct-integration-tct-only.

**Comment by plaidfinch** [comment]
> The `height` field is obviously wrong there, but there's also something wrong with the `Debug` implementation for `Root` that's causing it to print a truncated hash. Looking into the code, the `Root`'s `Debug` implementation is derived from that of `Hash`, which is doing something mysterious:
> 
> ```rust
> impl Debug for Hash {
>     fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
>         let mut bytes = Vec::with_capacity(4 * 8);
>         self.0.write(&mut bytes).unwrap();
>         write!(f, "{}", hex::encode(&bytes[3 * 8 + 4..]))
>     }
> }
> ```

This was useful for debugging purposes when I was developing the tree, but it's now been fixed to print the whole hash in the tct-integration-tct-only branch.

**Pull request closed without merging**
